### PR TITLE
* combine CPU, NET and RAM

### PR DESCRIPTION
**Pull request by olexiybuyanskyy**
## Change Description

phase 1 - combine CPU and NET. we split stake into equal amounts
phase 2 - combine RAM

- remove actions buyram, buyrambytes and sellram in system contract
- remove memory related functionality from cleos
- remove exchange rates related function
- amount of available memory calculated using
   **ram_bytes = max_ram_size / REM max supply * REM stake**

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: programs/cleos/main.cpp [hunk 1]
```diff
@@ -1003,10 +1001,8 @@ struct create_account_subcommand {
       createAccount->add_option("ActiveKey", active_key_str, localized("The active public key or permission level for the new account"));
 
       if (!simple) {
-         createAccount->add_option("--stake-net", stake_net,
-                                   (localized("The amount of tokens delegated for net bandwidth")))->required();
-         createAccount->add_option("--stake-cpu", stake_cpu,
-                                   (localized("The amount of tokens delegated for CPU bandwidth")))->required();
+         createAccount->add_option("--stake-quantity", stake_quantity,
```
Comment (roman-tik) [inline programs/cleos/main.cpp]: **--stake** should be fine. No need to add **-quantity** here

Comment (olexiybuyanskyy) [reply programs/cleos/main.cpp]: changed

File: contracts/contracts/eosio.system/src/exchange_state.cpp [hunk 2]
```diff
@@ -6,76 +6,6 @@
 
 namespace eosiosystem {
 
-   asset exchange_state::convert_to_exchange( connector& reserve, const asset& payment )
```
Comment (roman-tik) [inline contracts/contracts/eosio.system/src/exchange_state.cpp]: I propose to leave these methods here without actual use for now

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 3]
```diff
@@ -434,22 +295,25 @@ namespace eosiosystem {
...
 
    void system_contract::delegatebw( const name& from, const name& receiver,
-                                     const asset& stake_net_quantity,
-                                     const asset& stake_cpu_quantity, bool transfer )
+                                     const asset& stake_quantity,
+                                     bool transfer )
    {
       asset zero_asset( 0, core_symbol() );
+      const asset stake_net_quantity = stake_quantity / 2;
+      const asset stake_cpu_quantity = stake_quantity - stake_net_quantity;
```
Comment (roman-tik) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: where do you set ram quota?

File: contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp [hunk 4]
```diff
@@ -640,15 +610,14 @@ namespace eosiosystem {
           *    tokens to be staked,
           * @param receiver - the account to delegate bandwith to, that is, the account to
           *    whose resources staked tokens are added
-          * @param stake_net_quantity - tokens staked for NET bandwidth,
-          * @param stake_cpu_quantity - tokens staked for CPU bandwidth,
+          * @param stake_quantity - tokens staked for resources: NET and CPU,
```
Comment (roman-tik) [inline contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp]: + RAM

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp (hunk 3 above)
Comment (olexiybuyanskyy) [reply contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: we compute ram_bytes in
`changebw( from, receiver, stake_net_quantity, stake_cpu_quantity, transfer);`

File: contracts/contracts/eosio.system/src/exchange_state.cpp (hunk 2 above)
Comment (olexiybuyanskyy) [reply contracts/contracts/eosio.system/src/exchange_state.cpp]: ok

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp (hunk 3 above)
Comment (roman-tik) [reply contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: so, why don't we divide by 3 then?

Comment (olexiybuyanskyy) [reply contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: Because only the stake for CPU and NET used as a weight value. In general, we may change and set the stake to both NET and CPU without division. There is no difference.

Comment (olexiybuyanskyy) [reply contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: But if we would want to combine CPU and NET inside of **changebw** function it will require more changes in this function related to transfer 
```
auto transfer_amount = net_balance + cpu_balance;
         if ( 0 < transfer_amount.amount ) {
            token::transfer_action transfer_act{ token_account, { {source_stake_from, active_permission} } };
            transfer_act.send( source_stake_from, stake_account, asset(transfer_amount), "stake bandwidth" );
         }
```

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 5]
```diff
@@ -273,20 +126,26 @@ namespace eosiosystem {
...
          } else {
             totals_tbl.modify( tot_itr, from == receiver ? from : same_payer, [&]( auto& tot ) {
-                  tot.net_weight    += stake_net_delta;
-                  tot.cpu_weight    += stake_cpu_delta;
+                  tot.net_weight    += stake_delta;
+                  tot.cpu_weight    += stake_delta;
+                  tot.stake_amount  += stake_delta.amount;
                });
          }
          check( 0 <= tot_itr->net_weight.amount, "insufficient staked total net bandwidth" );
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp:147]: do we need this check?

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 6]
```diff
@@ -299,84 +158,66 @@ namespace eosiosystem {
                net_managed = has_field( voter_itr->flags1, voter_info::flags1_fields::net_managed );
                cpu_managed = has_field( voter_itr->flags1, voter_info::flags1_fields::cpu_managed );
             }
-
-            if( !(net_managed && cpu_managed) ) {
+            if( !(net_managed && cpu_managed && ram_managed) ) {
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp:162]: Do we need this if?

File: contracts/contracts/eosio.system/src/exchange_state.cpp [hunk 7]
```diff
@@ -28,7 +28,7 @@ namespace eosiosystem {
       const double Fi = double(1) / reserve.weight;
 
       double dR = R0 * ( std::pow(1. + dS / S0, Fi) - 1. ); // dR < 0 since dS < 0
-      if ( dR > 0 ) dR = 0; // rounding errors
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/exchange_state.cpp:31]: @olexiybuyanskyy please revert this unnecessary change

File: contracts/contracts/eosio.system/src/rex.cpp [hunk 8]
```diff
@@ -404,10 +404,6 @@ namespace eosiosystem {
                                  cpu_managed ? cpu : tot_itr->cpu_weight.amount );
          }
       }
-
-      if ( tot_itr->is_empty() ) {
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/rex.cpp:408]: @olexiybuyanskyy could you please explain this change?

File: contracts/contracts/eosio.token/include/eosio.token/eosio.token.hpp [hunk 9]
```diff
@@ -134,6 +134,21 @@ namespace eosio {
...
 
+         /**
+          * Get max supply method.
+          *
+          * @details Gets the max supply for token `sym_code`, created by `token_contract_account` account.
+          *
+          * @param token_contract_account - the account to get the supply for,
+          * @param sym_code - the symbol to get the supply for.
+          */
+         static asset get_max_supply( const name& token_contract_account, const symbol_code& sym_code )
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.token/include/eosio.token/eosio.token.hpp:145]: @olexiybuyanskyy what is purpose\use-case of this function?

File: contracts/contracts/eosio.system/src/eosio.system.cpp [hunk 10]
```diff
@@ -72,6 +72,15 @@ namespace eosiosystem {
       _global4.set( _gstate4, _self );
    }
 
+   void system_contract::setminstake( uint64_t min_account_stake ) {
+      require_auth( _self );
+
+      // 1,000,000'0000 REM is not realistic min stake
+      check( min_account_stake < 10000000000, "min stake is unrealistic" );
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/eosio.system.cpp]: @olexiybuyanskyy do we really need this check? if so maybe we can associate it with `_gstate.total_activated_stake` e.g. not more then 1% of total activate stake

**Review by mark-tsyrulnyk — APPROVED** [review]

File: contracts/contracts/eosio.token/include/eosio.token/eosio.token.hpp (hunk 9 above)
Comment (olexiybuyanskyy) [reply contracts/contracts/eosio.token/include/eosio.token/eosio.token.hpp:145]: used in changebw logic
```
const auto system_token_max_supply = eosio::token::get_max_supply(token_account, system_contract::get_core_symbol().code() );
               const double bytes_per_token = (double)_gstate.max_ram_size / (double)system_token_max_supply.amount;
               const int64_t bytes_for_stake = bytes_per_token * tot_itr->own_stake_amount;
```

File: contracts/contracts/eosio.system/src/voting.cpp [hunk 11]
```diff
@@ -119,11 +119,14 @@ namespace eosiosystem {
    double stake2vote( int64_t staked, time_point vote_mature_time ) {
       check(vote_mature_time != time_point(), "vote should have mature time");
 
-      const auto useconds_to_mature = fmax((vote_mature_time - current_time_point()).count(), 0.0);
-      const auto rem_weight = 1.0 - useconds_to_mature / eosio::days( system_contract::vote_mature_period ).count();
+      const auto seconds_to_mature = std::max( (vote_mature_time - current_time_point()).to_seconds(), 0LL );
+      const auto rem_weight = 1.0 - seconds_to_mature / eosio::days( system_contract::vote_mature_period ).to_seconds();
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/voting.cpp:123]: seconds_to_mature / eosio::days( system_contract::vote_mature_period ).to_seconds() is integer so rem_weight always will be either 0 or 1

File: contracts/contracts/eosio.system/src/eosio.system.cpp [hunk 12]
```diff
@@ -478,13 +351,13 @@ EOSIO_DISPATCH( eosiosystem::system_contract,
      // native.hpp (newaccount definition is actually in eosio.system.cpp)
      (newaccount)(updateauth)(deleteauth)(linkauth)(unlinkauth)(canceldelay)(onerror)(setabi)
      // eosio.system.cpp
-     (init)(setram)(setramrate)(setparams)(setpriv)(setalimits)(setacctram)(setacctnet)(setacctcpu)(activate)
+     (init)(setram)(setramrate)(setparams)(setpriv)(setalimits)(activate)
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/eosio.system.cpp:354]: need to add (setminstake) here to make action visible:
(init)(setram)(setminstake)(setramrate)(setparams)(setpriv)(setalimits)(activate)
